### PR TITLE
PoC retry transaction 2.1

### PR DESCRIPTION
**Pull request by akirayamamoto**
This PR aims to fix the intermittent test failures on
* https://github.com/OctopusDeploy/Nevermore/pull/235

### CODE REVIEW

File: source/Nevermore.IntegrationTests/Chaos/ChaosSqlCommand.cs [hunk 1]
```diff
@@ -20,7 +21,12 @@ public ChaosSqlCommand(DbCommand wrappedCommand, double chaosFactor)
         void MakeSomeChaos()
```
Comment (akirayamamoto) [inline source/Nevermore.IntegrationTests/Chaos/ChaosSqlCommand.cs:21]: `MakeSomeChaos` is working, this is what was breaking this branch: `core/poc-retry-transaction2`

File: source/Nevermore/RelationalStore.cs [hunk 2]
```diff
@@ -18,17 +17,6 @@ public class RelationalStore : IRelationalStore
         readonly Lazy<RelationalTransactionRegistry> registry;
         readonly Lazy<KeyAllocator> keyAllocator;
 
-        public static void EnableChaos(int percentChanceOfChaos)
```
Comment (akirayamamoto) [inline source/Nevermore/RelationalStore.cs:21]: I believe we can use `ChaosSqlCommand`.

File: source/Nevermore/CommandExecutor.cs [hunk 3]
```diff
@@ -54,7 +45,7 @@ public int ExecuteNonQuery()
             try
             {
                 AssertSynchronousOperation();
-                return command.ExecuteNonQueryWithRetry(retryPolicy, replayClosure);
+                return command.ExecuteNonQuery();
```
Comment (akirayamamoto) [inline source/Nevermore/CommandExecutor.cs:48]: Retry logic is now on the `ReadTransaction` class.

File: source/Nevermore.IntegrationTests/Chaos/ChaosSqlCommand.cs [hunk 4]
```diff
@@ -20,7 +21,12 @@ public ChaosSqlCommand(DbCommand wrappedCommand, double chaosFactor)
...
         {
             if (Debugger.IsAttached) return; // No chaos when debugging thanks!
-            if (ChaosGenerator.NextDouble() < chaosFactor) throw new TimeoutException("You made the chaos monkey angry...");
+            if (ChaosGenerator.NextDouble() < chaosFactor)
+            {
+                if (ChaosGenerator.NextDouble() < closeConnectionChaosFactor)
+                    wrappedCommand.Connection?.Close();
```
Comment (akirayamamoto) [inline source/Nevermore.IntegrationTests/Chaos/ChaosSqlCommand.cs:27]: When the chaos monkey kicks in there is a 50% of chance of dropping the connection.

File: source/Nevermore/Advanced/WriteTransaction.cs [hunk 5]
```diff
@@ -282,20 +282,24 @@ public void Commit()
         {
             if (Transaction is null)
                 throw new InvalidOperationException("There is no current transaction, call Open/OpenAsync to start a transaction");
+
             if (!configuration.AllowSynchronousOperations)
                 throw new SynchronousOperationsDisabledException();
 
             configuration.Hooks.BeforeCommit(this);
             Transaction.Commit();
+            ExecutedCommands.Clear();
```
Comment (akirayamamoto) [inline source/Nevermore/Advanced/WriteTransaction.cs:291]: Once the transaction is committed we need to clear the list of executed commands.

File: source/Nevermore/Advanced/ReadTransaction.cs [hunk 6]
```diff
@@ -651,9 +652,17 @@ public Task<int> ExecuteNonQueryAsync(string query,  CommandParameterValues? arg
         public int ExecuteNonQuery(PreparedCommand preparedCommand)
         {
             using var mutex = DeadlockAwareLock.Lock();
-            using var command = CreateCommand(preparedCommand);
-            var result =  command.ExecuteNonQuery();
+            var retryPolicy = GetRetryPolicy(preparedCommand.Operation).LoggingRetries(nameof(ExecuteNonQuery));
+            var result = retryPolicy.ExecuteAction(() =>
+            {
+                if (connection?.State is ConnectionState.Closed)
```
Comment (akirayamamoto) [inline source/Nevermore/Advanced/ReadTransaction.cs:658]: Is this enough to determine that we need to replay the transaction?

Comment (stevencl840) [reply source/Nevermore/Advanced/ReadTransaction.cs:658]: I am not sure to be honest. Should it be using the IsTransient check? Is that possible here?

Comment (michaelongso) [reply source/Nevermore/Advanced/ReadTransaction.cs:658]: The "replay" state should be determined when detecting the transient error

**Review by stevencl840 — APPROVED** [review]
Probably needs more work around error detecting but i think this is an improvement to the POC

**Pull request merged**